### PR TITLE
Fixes #4399 - Spacemacs documentation improvements and fixes.

### LINKED ISSUE #4399: Broken links in some .org files
The TOC links in the following files are broken (404):

* https://github.com/syl20bnr/spacemacs/blob/master/doc/DOCUMENTATION.org
* https://github.com/syl20bnr/spacemacs/blob/master/doc/FAQ.org
* https://github.com/syl20bnr/spacemacs/blob/master/doc/CONVENTIONS.org
* https://github.com/syl20bnr/spacemacs/blob/master/doc/LAYERS.org
* https://github.com/syl20bnr/spacemacs/blob/master/doc/QUICK_START.org
* https://github.com/syl20bnr/spacemacs/blob/master/doc/VIMUSERS.org
* https://github.com/syl20bnr/spacemacs/blob/master/layers/LAYERS.org
* layers/**/README.org (I haven't tried all of them)

I think the broken links were introduced in 13c5b1d (Nov 2, 2015).

PS: the TOC links in https://github.com/syl20bnr/spacemacs/blob/master/CONTRIBUTING.org are not broken.

-----
**Update:** Not only TOC links, but most (if not all) internal links.

### PR DESCRIPTION
**Pull request by JAremko**
The prime goal of this PR is to fix the Spacemacs documentation links in the Table Of Content at GitHub and introduce reliable method of linking headings from different files that will work in the Spacemacs and both at GitHub and spacemacs.org web sites. 

**I fixed all the merge conflict except the last commit (it's too global and will constantly brake)**

 In the discussion  #4399  we came to the conclusion that the solution shouldn't modify local files in the Spacemacs installation - they should be the same locally and at the repository.  Also we decided  to use toc-org for table of content.

Since  we can't change how GitHub handles links, I made Spacemacs and spacemacs.org  use the GitHub style links - they should always work at GitHub. Also I added some of my PRs that overlaps with this one and smashed bugs, fixed typos that I encountered while implementing the fix. They all are in the separate commits so it shouldn't be a problem. 

### I'll explain each of the commits
 

 - **fix typo	3b5ffac**   it fixes typo.  

---

 - **Remove `query.stickytableheaders.js` (error 404)	ebad4eb**  it causes 404 error at [the Spacemacs website](http://spacemacs.org/layers/LAYERS.html)(open web-browser console to see it) The  file [isn't there anymore](http://www.pirilampo.org/styles/lib/js/jquery.stickytableheaders.js), but it's not crucial and the website should be fine without it.

---

 - **Insert the readtheorg.css links wth html export advice 	2012530**  Many README.org files of the Spacemacs layers have wrong links to the `readtheorg.css`  file. Also it adds unnecessary clutter   The commit adds or fixes (if it's already there) the link during the export to html process. Also it implements a wrapper function that can chain all kind of the org file preprocessors (used later)  

---

 - **add TOC to Semantic layer	bfe1d73**  it was missing. Now it's not.

---

 - **Make export to html compatible with GitHub and toc-org	d0abe2e** this commit adds the export preprocessors:
    - `spacemacs//toc-org-unhrefify-toc` - makes TOC_gh normal org TOC again.
    - `spacemacs//reroot-links` - If a link refers an org file or its heading and the file is in the Spacemacs GitHub repository, the function will rewrite the base of the link to http://spacemacs.org/  | [example](http://jaremko.github.io/spacemacs.org/doc/pr-test/test.html)  |  [exported org file](https://raw.githubusercontent.com/syl20bnr/spacemacs/1706d3fe36e8360930a69b13191a4d08e18c73e3/doc/pr-test/test.org). 
    - `spacemacs//org-heading-annotate-custom-id` - adds GitHub style ids to the exported headings so they can be referred in the same way as at the GitHub web site.  *For the more info read  the commit message.*

---

 - **Add Spacemacs docs minor mode to hide meta tags. 	b8f0ea7**   When the mode is enabled, it hides nonsense in the docs:
    - **Off (the image is clickable)** 
[](http://i.imgur.com/id2uPNu.png)
    - **On (the image is clickable)** 
[](http://i.imgur.com/6oZzWz4.png)

---

 - **Add org-mode link-type "https" to open local copies  	59ae423**  this is an `org-mode` `https` link type handler that brings GitHub links to the Spacemacs(including links to headings) In a similar fashion to `spacemacs//reroot-links`  if a link refers an org file or its heading and the file is in the Spacemacs GitHub repository, the handler will open the local copy of the file with `spacemacs/view-org-file`(this is how `SPC` `h` `SPC` opens them) or if it's not in there - the link will be opened in the browser. The function `spacemacs/view-org-file` applies visual enhancements like the indent mode and `space-doc-mode` so the opened files also will be prettified. And since those are the GItHub style links they will work at the GitHub, Spacemcs and spacemacs.org. For example: 
    - `[[https://github.com/syl20bnr/spacemacs/blob/master/doc/FAQ.org#os-x][Link to FAQ "OS X" heading]]`  leads to **(the image is clickable)**  [](http://i.imgur.com/5Pf9UUZ.png) 
    -  but `[[https://www.google.com][Link to www.google.com]]` leads to https://www.google.com  
This is more reliable and flexible way to link files and headings than the default `org-mode` one - if we want it to work at the GitHub, in the Spacemacs and at its web site. + it adds pretifiers. 

---

 - **Remove #+HTML_HEAD_EXTRA: readtheorg.css 	100bc0a**  now we don't need them.

---

 - **Set org files TOC to :TOC_4_gh:noexport:	88e5ff9** This way `toc-org` will build GitHub TOC when the org files are saved.

---

 - **~PR TEST COMMIT~ 	1706d3f**  `toc-org-gh.el`  script for the applying `toc-org` to all the org files in the batch mode and example files for the PR.  You can call the script from the repo root with `find . -name "*.org" -type f -exec emacs -batch -l ./doc/pr-test/toc-org-gh.el '{}' -f toc-apply \;`

---

 -  **format .org TOCs with toc-org	402bba5** Contains org files processed with `toc-org-gh.el` - not for merge. Can be easily reproduced by running the script.  *Probably should be split into a couple of commits, because it makes git crush - too many changes?*

#### @syl20bnr  @TheBB  @StreakyCobra   @a13ph @robbyoconnor  @d12frosted and everyone else! Be the witnesses of my glory or the miserable failure!

### CODE REVIEW

**Comment by JAremko** [comment]
We should discuss it. And if this is a proper way to go around I'll write about the links in the Contributing guidelines and add .org validations to the test suite.

Also most of the commits can be safely cherry picked.

File: core/core-documentation.el [hunk 1]
```diff
@@ -85,19 +88,99 @@
...
+  (save-current-buffer
+    (save-excursion
+      (let* ((filename (car (nthcdr 1 args)))
+             (visitingp (find-buffer-visiting filename)))
+        ;; Temporary "unvisit" the visited org files.
+        (when visitingp (with-current-buffer visitingp (setq buffer-file-name nil)))
+        (with-temp-buffer
+          (save-match-data
+            (insert-file-contents filename t)
+            (spacemacs//add-org-meta-readtheorg-css filename)
```
Comment (syl20bnr) [inline core/core-documentation.el]: You can add a comment here to specify that any preprocess function should be added here.

**Comment by syl20bnr** [comment]
A quick look at the PR tells me that it is pretty much what I wanted, nice work :+1: 
Kudos for the script!

**Comment by robbyoconnor** [comment]
Merge conflicts :P

**Comment by JAremko** [comment]
@robbyoconnor  Ya it's the last commit.  It should be ditched anyway - I have no hope to keep it in sync :smile:

**Comment by robbyoconnor** [comment]
Not with that attitude you don't!

**Comment by JAremko** [comment]
@syl20bnr  I removed commits that change .org files and improved the formatting script. 
Now you just need to run `bash ./doc-fmt/run.bash` and it will:
  - remove  `#+HTML_HEAD_EXTRA ... readtheorg.css" />`
  - replace :TOC_4_org: with :TOC_4_gh:
  - apply toc-org 

This way the PR shouldn't get conflicts so often.

**Comment by syl20bnr** [comment]
Great move!

mrclean

**Comment by robbyoconnor** [comment]
I live this.

--Rob

File: core/core-documentation.el (hunk 1 above)
Comment (JAremko) [inline core/core-documentation.el]: done.

**Comment by JAremko** [comment]
woops :sleeping:

**Comment by JAremko** [comment]
Dropped `add TOC to Semantic layer bfe1d73` because the layer was nuked or moved in develop.

**Comment by JAremko** [comment]
updated update README.org.template  4c521bb

**Comment by syl20bnr** [comment]
One of my favorite PR, clean and fun to merge. Great job :purple_heart: 
I ran the script on the doc files and it worked great. Also the new preprocessors fixed some wrong CSS paths :+1: 
So this is it, we have links working in Emacs and Github! :tada: 

jaremko

Thank you !
Cherry-picked into develop branch, you can safely delete your branch.

**Comment by JAremko** [comment]
Thanks! :blush:

**Comment by robbyoconnor** [comment]
I love you all

**Pull request closed without merging**